### DISCUSSION TITLE
HTTP와 HTTPS의 차이를 설명하고 HTTPS 통신 원리에 대해 설명해주세요.

### DISCUSSION THREAD

**Discussion by nak-honest** [woowacourse-6th-computer-sign-pen/comsa-interview-study#12] [Network] [answered]
HTTP와 HTTPS의 차이를 설명하고 HTTPS 통신 원리에 대해 설명해주세요.

**Comment by slimsha2dy** [comment, marked as answer]
**답변자: 낙낙(이낙헌)**

질문자: HTTP가 무엇인가요?

답변자: HTTP는 HyperText Transfer Protocol의 약자로, 애플리케이션 계층의 프로토콜입니다. 네트워크 통신 시 하이퍼텍스트를 전송할 때 사용됩니다.

질문자: 장단점은 무엇인가요?

답변자: 장점은 HTTPS에 비해 암호화 과정이 없어서 속도가 빠릅니다. 하지만 단점은 데이터가 암호화되지 않기 때문에 보안에 취약하다는 것입니다. 중간에서 패킷을 가로채면 내용이 그대로 노출될 수 있습니다.

질문자: 보안에 취약하다는 것은 구체적으로 어떤 의미인가요?

답변자: 네트워크 패킷을 중간에서 가로채면 암호화되지 않은 모든 내용을 볼 수 있습니다. 예를 들어, 아이디나 비밀번호 같은 민감한 정보가 노출될 수 있어요.

질문자: 그렇다면 HTTPS는 무엇인가요?

답변자: HTTPS는 HTTP에 SSL/TLS 프로토콜을 추가한 것으로, 데이터 통신을 암호화하여 보안을 강화한 프로토콜입니다.

질문자: 암호화는 어떻게 이루어지나요?

답변자: 먼저, 클라이언트와 서버 간에 SSL/TLS 핸드셰이크 과정을 거칩니다.
클라이언트가 서버에 접속 요청을 보냅니다. 이때 사용 가능한 암호화 알고리즘 목록을 함께 보냅니다.
서버는 암호화 알고리즘을 선택하고, SSL 인증서를 클라이언트에게 제공합니다. 이 인증서에는 서버의 공개 키와 서버의 신원을 증명하는 정보가 포함되어 있습니다.
클라이언트는 인증서를 검증합니다. 인증서를 발급한 CA의 공개 키를 사용하여 인증서의 유효성을 확인합니다.
클라이언트는 대칭 키를 생성하고, 서버의 공개 키로 암호화하여 서버에 전송합니다.
서버는 자신의 개인 키로 암호화된 대칭 키를 복호화합니다. 이제 클라이언트와 서버는 동일한 대칭 키를 공유하게 되며, 이 키를 사용하여 안전한 통신을 진행합니다.

질문자: 서버가 제공하는 인증서는 누구의 것인가요?

답변자: 서버의 인증서입니다. 신뢰할 수 있는 인증 기관(CA)이 발급한 것으로, 서버의 신원을 보증하고 공개 키를 포함하고 있습니다.

질문자: 클라이언트는 그 인증서가 진짜인지 어떻게 알 수 있나요?

답변자: 클라이언트는 CA의 공개 키를 사용하여 인증서를 검증합니다. 브라우저나 운영체제에는 신뢰할 수 있는 루트 CA들의 공개 키가 내장되어 있습니다. 인증서의 디지털 서명을 검증함으로써 인증서의 진위 여부를 확인합니다.

질문자: SSL과 TLS의 차이는 무엇인가요?

답변자: SSL은 Secure Sockets Layer의 약자이고, TLS는 Transport Layer Security의 약자입니다. SSL이 먼저 개발되었고, 이후에 보안 강화와 알고리즘 개선을 통해 TLS가 나왔습니다. 현재는 TLS가 표준이지만, 관습적으로 SSL이라고 부르는 경우도 많습니다.

질문자: 대칭 키 방식과 공개 키 방식의 장단점은 무엇인가요?

답변자:
대칭 키 암호화: 하나의 키로 암호화와 복호화를 모두 수행하므로 속도가 빠릅니다. 하지만 키를 공유해야 하는 문제(키 분배 문제)가 있어서 안전하게 키를 전달하기 어렵습니다.
공개 키 암호화(비대칭 키 암호화): 공개 키와 개인 키 한 쌍을 사용하여 키 분배 문제를 해결합니다. 그러나 연산 속도가 느려서 대용량 데이터 암호화에는 부적합합니다.

질문자: 키 분배 문제를 해결할 다른 방법이 있나요?

답변자: 네, 디피-헬만 키 교환 알고리즘이 있습니다. 이는 두 당사자가 비밀 키를 공유하지 않고도 안전하게 공통의 대칭 키를 생성할 수 있는 방법입니다. 하지만 중간자 공격에 취약할 수 있어, 일반적으로는 인증서를 통한 신원 확인과 함께 사용됩니다.

질문자: CSRF 공격에 대해 아시나요?

답변자: 자세히 알지는 못하지만, CSRF(Cross-Site Request Forgery)는 사용자가 인증된 상태를 악용하여 공격자가 의도한 행동을 수행하게 만드는 웹 보안 공격입니다.

질문자: 디지털 서명은 어떻게 동작하나요?

답변자: 디지털 서명은 송신자의 개인 키로 데이터를 암호화하는 방식입니다. 수신자가 송신자의 공개 키로 복호화를 성공하면, 이는 송신자의 개인 키로 암호화가 되었다는 이야기인데, 이말인즉슨 송신자의 개인키로 암호화 했다는 것을 증명하는 셈이 됩니다. 이를 통해 데이터가 송신자로부터 왔으며 중간에 변조되지 않았음을 보장합니다.

질문자: 설명 감사합니다.

답변자: 추가로, 디피-헬만 알고리즘은 두 당사자가 안전하게 비밀 키를 생성하는 방법으로, 수학적인 원리를 활용합니다. 그림으로 표현하면 더 이해하기 쉬운데, 색을 섞는 비유를 자주 사용합니다.

질문자: 오늘 인터뷰는 여기까지 하겠습니다. 수고하셨습니다.
